### ISSUE TITLE
Proposal: Adding AUTO_INCREMENT Feature to ClickHouse

### ISSUE THREAD

**Issue by yusufozturk** [ClickHouse/ClickHouse#56228]
PS: A quick proposal after our talk in Amsterdam meetup @alexey-milovidov 
Some of the technical details are beyond my expertise, so I had to ask help from GPT to write this proposal

---

### Introduction

The utility of an AUTO_INCREMENT feature, which generates unique integer values for specific columns, is evident in numerous database systems. The addition of this capability to ClickHouse can enhance data management, facilitate optimized querying, and provide a consistent method for row identification.

---

### Design Considerations

1. **Implementation Model**: 
    - Given ClickHouse's distributed architecture, adopting a distributed unique ID generation system, akin to Twitter's Snowflake, is recommended. This ensures unique IDs across nodes, a crucial aspect for distributed databases like ClickHouse.

2. **DataType**:
    - The column with AUTO_INCREMENT should have the `Int64` data type, which is ClickHouse's equivalent of `BIGINT` in other databases.

3. **Uniqueness and Sequencing**:
    - While AUTO_INCREMENT will ensure the uniqueness of values, it doesn't need to guarantee:
        - Consecutive or sequential ordering.
        - Starting from a specific value.
        - Distinction from manually set values.

4. **Constraints**:
    - **Unique Identifier**: While ClickHouse possesses the concept of a primary key, it's primarily for optimizing data retrieval rather than ensuring uniqueness. The AUTO_INCREMENT feature would emphasize providing a unique identifier for rows, rather than acting as a traditional primary key.
    - Each table can have only one AUTO_INCREMENT column.
    - The AUTO_INCREMENT column should be NOT NULL and should not have a default value.

5. **Value Assignment**:
    - When a new row is inserted without a value for the AUTO_INCREMENT column, ClickHouse should auto-assign a value.
    - Explicitly set values during INSERT or UPDATE should be accepted, but the system must ensure that future auto-generated values avoid collisions with these.

6. **Management Across Nodes**:
    - Given ClickHouse's distributed environment, each node might need to manage a specific range of AUTO_INCREMENT values, thus preventing collisions. This would require further technical examination to align with ClickHouse's architecture.

7. **Alternative for Custom Sequencing**:
    - If the default AUTO_INCREMENT behavior isn't suitable for specific use cases, we could consider offering users a custom sequencing mechanism.

---

### Benefits of AUTO_INCREMENT in ClickHouse

1. **Unique Row Identification**: Provides a consistent method for uniquely identifying rows, improving data manageability.
2. **Optimized Querying**: The presence of a unique identifier can facilitate more efficient data retrieval operations.
3. **Consistent Data Management**: A standardized way to handle row identification can streamline operations and reduce complexities.

---

### Challenges & Considerations

1. **Managing Explicit Values**: Ensuring no collisions when users manually set a value in the AUTO_INCREMENT column.
2. **Distributed System Nuances**: Given ClickHouse's distributed nature, managing AUTO_INCREMENT values across nodes is challenging and would require a tailored approach.

---

### Conclusion

Incorporating the AUTO_INCREMENT feature into ClickHouse can bring about enhanced usability and performance for various data management scenarios. While there are specific nuances and challenges to address, a thoughtfully designed implementation can make this a valuable addition to ClickHouse's suite of features.

**Comment by alexey-milovidov** [comment]
I propose this implementation:

A new function, `serial`, which takes a constant string argument - a series identifier.
On every block, it goes to Keeper and atomically increments the counter with the corresponding name by the block size, then it generates and returns sequential numbers starting from the previous counter value.

The server can be configured with a keeper path - a prefix for these counter nodes, and with a limit on the total number of distinct counters (e.g., 100,000).

The usage of `serial` function will be controlled by a separate access right.

There should be a shortcut syntax for a table definition: if you specify `AUTO_INCREMENT`, it is equivalent to `DEFAULT serial('`db`.`table`')`

As a result, it will provide a monotonic, but not necessarily continuous counter. The discontinuity will appear for rejected or deduplicated INSERTs.

**Comment by den-crane** [comment]
Other databases call such functions `sequence`

https://www.postgresql.org/docs/current/sql-createsequence.html

https://docs.oracle.com/cd/B12037_01/server.101/b10759/statements_6014.htm

```
CREATE SEQUENCE customers_seq
 START WITH     1000
 INCREMENT BY   1
 CACHE 100000000 -- increment in the Zookeeper by 100000000 
;
```

**Comment by den-crane** [comment]
past discussion
* https://github.com/ClickHouse/ClickHouse/issues/5681

**Comment by yusufozturk** [comment]
@alexey-milovidov I like the idea but I am a bit concerned about insert performance. Using keeper for this operation might bring some additional cost for high-frequency insert operations or batch inserts. Also single node installation would require different implementation, which would bring complexity on documentation and UX.

I am a simple minded guy so I'm still keen on snowflakeid idea. Because I feel like that's the easiest and least painful way. Even in a really large Clickhouse deployment, with the proper implementation and assuming no clock skew, the collision rate should be zero with snowflakeid. But you are the master Jedi. Maybe I don't see all user cases properly.

**Comment by alexey-milovidov** [comment]
@yusufozturk Yes, Twitter's snowflake id will be much easier to implement.

**Comment by UnamedRus** [comment]
> Maybe I don't see all user cases properly.

Probably, it's not about current version of ClickHouse, but it's really nice to have not sparse increment column, because it fits really nice in memory structure like FLAT dictionary (basically array with small percent of holes).

**Comment by Slach** [comment]
@yusufozturk 
Snowflake id already can be easy implements 

```sql
CREATE FUNCTION toSnowflake64 as (dt,ch) ->
  bitOr(dateTime64ToSnowflake(dt),
   bitAnd(bitAnd(ch,0x3FFFFF)+
      bitAnd(bitShiftRight(ch, 20),0x3FFFFF)+
      bitAnd(bitShiftRight(ch, 40),0x3FFFFF),
      0x3FFFFF) 
  );

CREATE FUNCTION toSnowflake as (dt,ch) ->
  bitOr(dateTimeToSnowflake(dt),
   bitAnd(bitAnd(ch,0xFFFFFFFF)+
      bitAnd(bitShiftRight(ch, 32),0xFFFFFFFF),
      0xFFFFFFFF) 
  );
    

CREATE TABLE t (id UInt64 DEFAULT toSnowflake64(now64(), hostName()), s LowCardinality(String)) ENGINE=MergeTree() ORDER BY tuple();

```

**Comment by yusufozturk** [comment]
@Slach Thanks a lot Eugene. We will try this ASAP :)

**Comment by Slach** [comment]
@yusufozturk this is not auto-increment actually this will get the same value for whole insert block.

More randomized and not auto-increment
```sql
CREATE TABLE t (
 id UInt64 DEFAULT toSnowflake64(now64(), cityHash64(concat(hostName(),toString(rowNumberInAllBlocks())))), 
 s LowCardinality(String)
) ENGINE=MergeTree() ORDER BY tuple();

INSERT INTO t(s) SELECT toString(number) FROM numbers(1000);

SELECT * FROM t;
```

and much slower
```
 id UInt128 DEFAULT toUInt128(concat(toString(toSnowflake64(now64(), cityHash64(hostName()))),leftPad(toString(rowNumberInAllBlocks()), 9, '0'))), 
cityHash64(hostName()))),toString(rowNumberInAllBlocks())))), 
```

https://fiddle.clickhouse.com/7b9da9c0-afdb-49f9-bc10-03ba8706f558
auto increment usually have no sense for analytics.

**Comment by CH2Dev** [comment]
@Slach @clickhouse-admin 

I want function query for dateTimeToSnowflake & dateTime64ToSnowflake........ Clickhouse Create query & .Net Code........ This will be really helpful if you share it ASAP

**Comment by yusufozturk** [comment]
Thanks for the 24.6:
https://x.com/ClickHouseDB/status/1808187054070669340

**Comment by flyly0755** [comment]
@yusufozturk Your guys think too much about distributed architecture, but in lots of scenerios,  we just need a non-distributed table with auto-increment feature.  Even with distributed scenerio, usually with a non-distributed main table and distributed

**Comment by alexey-milovidov** [comment]
@flyly0755, implemented in 25.1, https://presentations.clickhouse.com/2025-release-25.1/index.html
